### PR TITLE
Make IndexShard operation be more explicit about whether they are expected to run on a primary or replica (backport #15282)

### PR DESCRIPTION
**Pull request by bleskes**
This commit cherry picks some infrastructure changes from the `feature/seq_no` branch to make merging from master easier.

More explicitly, IndexShard current have prepareIndex and prepareDelete methods that are called both on the primary as the replica, giving it a different origin parameter. Instead, this commits creates two explicit prepareOnPrimary and prepareOnReplica methods. This has the extra added value of not expecting the caller to use an Engine enum.

Also, the commit adds some code reuse between TransportIndexAction and TransportDeleteAction and their TransportShardBulkAction counter parts.

This is a backport of #15282

### CODE REVIEW

**Comment by bleskes** [comment]
@jasontedor this was a non-trivial back port. Would love a sanity check. Thx.

File: core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java [hunk 1]
```diff
@@ -178,35 +178,83 @@ protected ShardIterator shards(ClusterState clusterState, InternalRequest reques
...
+        Mapping update = operation.parsedDoc().dynamicMappingsUpdate();
+        final ShardId shardId = indexShard.shardId();
+        if (update != null) {
+            final String indexName = shardId.getIndex();
+            mappingUpdatedAction.updateMappingOnMasterSynchronously(indexName, request.type(), update);
+            operation = prepareIndexOperationOnPrimary(shardRequest, request, indexShard);
+            update = operation.parsedDoc().dynamicMappingsUpdate();
+            if (update != null) {
+                throw new RetryOnPrimaryException(shardId,
+                    "Dynamics mappings are not available on the node that holds the primary yet");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java]: Typo: "Dynamics" -> "Dynamic"

**Comment by jasontedor** [comment]
Saw a minor typo, looked carefully over the places where the code being changed deviates substantially from master but it looks like you got everything. LGTM.

**Comment by bleskes** [comment]
pushed to 2.x Thanks @jasontedor

**Pull request closed without merging**